near-contract-standards/src/fungible_token/implementation.rs

```diff
@@ -0,0 +1,256 @@
+            sender_id,
+            receiver_id.into(),
+            amount.into(),
+            &env::current_account_id(),
+            NO_DEPOSIT,
+            GAS_FOR_RESOLVE_TRANSFER,
+        ))
+    }
+
+    fn ft_total_supply(&self) -> U128 {
+        self.total_supply.into()
+    }
+
+    fn ft_balance_of(&self, account_id: ValidAccountId) -> U128 {
+        self.accounts.get(account_id.as_ref()).unwrap_or(0).into()
+    }
+}
+
+impl FungibleTokenResolver for FungibleToken {
+    fn ft_resolve_transfer(
+        &mut self,
+        sender_id: ValidAccountId,
+        receiver_id: ValidAccountId,
+        amount: U128,
+    ) -> U128 {
+        let amount: Balance = amount.into();
+        let sender_id: AccountId = sender_id.into();
+        let receiver_id: AccountId = receiver_id.into();
+
+        // Get the unused amount from the `ft_on_transfer` call result.
+        let unused_amount = match env::promise_result(0) {
+            PromiseResult::NotReady => unreachable!(),
+            PromiseResult::Successful(value) => {
+                if let Ok(unused_amount) = near_sdk::serde_json::from_slice::<U128>(&value) {
+                    std::cmp::min(amount, unused_amount.0)
+                } else {
+                    amount
+                }
+            }
+            PromiseResult::Failed => amount,
+        };
+
+        if unused_amount > 0 {
+            let receiver_balance = self.accounts.get(&receiver_id).unwrap_or(0);
+            if receiver_balance > 0 {
+                let refund_amount = std::cmp::min(receiver_balance, unused_amount);
+                self.accounts.insert(&receiver_id, &(receiver_balance - refund_amount));
+
+                if let Some(sender_balance) = self.accounts.get(&sender_id) {
+                    self.accounts.insert(&sender_id, &(sender_balance + refund_amount));
+                    log!("Refund {} from {} to {}", refund_amount, receiver_id, sender_id);
+                    return (amount - refund_amount).into();
+                } else {
+                    // Sender's account was deleted, so we need to burn tokens.
+                    self.total_supply -= refund_amount;
+                    log!("The account of the sender was deleted");
+                    log!("Burn {}", refund_amount);
+                }
+            }
+        }
+        amount.into()
+    }
+}
+
+impl StorageManager for FungibleToken {
+    fn storage_deposit(&mut self, account_id: Option<ValidAccountId>) -> AccountStorageBalance {
+        let amount = env::attached_deposit();
+        assert_eq!(
+            amount,
+            self.storage_minimum_balance().0,
+            "Requires attached deposit of the exact storage minimum balance"
+        );
+        let account_id =
+            account_id.map(|a| a.into()).unwrap_or_else(|| env::predecessor_account_id());
+        self.internal_register_account(&account_id);
+        AccountStorageBalance { total: amount.into(), available: amount.into() }
```

Comment:
Also want to point out that there's a bit of a confusion on this.
Al believes this should essentially return `available: 0` and believes this method is separate from account registration. He voiced it here: https://github.com/near/NEPs/pull/158#discussion_r582090178
Looking at this method and `storage_withdraw` (where we have `…self.accounts.remove(…`) it's pretty clear these are account registration and unregistration. Just want to connect dots here, as I don't think we've brought this up with Eugene directly.

Reply:
Let's discuss on the meeting tomorrow.